rules/import_middleman.bzl

```diff
@@ -0,0 +1,227 @@
+        "force_load_library",
+        "source",
+        "link_inputs",
+        "linkopt",
+        "library",
+        "dynamic_framework_file",
+        "static_framework_file",
+    ]:
+        set = depset(
+            direct = [],
+            # Note:  we may want to merge this with the below inputs?
+            transitive = [getattr(dep[apple_common.Objc], key) for dep in ctx.attr.deps],
+        )
+        _add_to_dict_if_present(objc_provider_fields, key, set)
+
+    # Build out the imported libraries replacing each one along the way
+    filtered_libs = []
+    existing_libs = {}
+    for li in input_imported_libraries:
+        existing_libs[li.path] = True
+        filtered_libs.append(_do_lib(ctx, li))
+
+    exisiting_imported_libraries = objc_provider_fields.get("imported_library", depset([]))
+    for li in exisiting_imported_libraries.to_list():
+        if not existing_libs.get(li.path, False):
+            filtered_libs.append(li)
+
+    objc_provider_fields["imported_library"] = depset(filtered_libs)
+
+    # Build out framework inputs
+    # FIXME: this should follow the above logic, or simplify
+    static_framework_file = [_do_framework(ctx, f) for f in input_static_frameworks]
+    dynamic_framework_file = []
+    dynamic_framework_dirs = []
+    for f in input_dynamic_frameworks:
+        out = _do_framework(ctx, f)
+        dynamic_framework_file.append(out)
+        dynamic_framework_dirs.append(out)
+
+        # Append ad-hoc framework files by name: e.g. ( Info.plist )
+        ad_hoc_file = all_import_infos[f.path].framework_imports.to_list()
+
+        # Remove the input framework files from this rule
+        ad_hoc_file.remove(f)
+        dynamic_framework_dirs.extend(ad_hoc_file)
+
+    all_frameworks = static_framework_file + dynamic_framework_file
+
+    # TODO: see above comment
+    objc_provider_fields["linkopt"] = depset(
+        ["\"-F" + "/".join(f.path.split("/")[:-2]) + "\"" for f in static_framework_file],
+        transitive = [
+            objc_provider_fields.get("linkopt", depset([])),
+        ],
+    )
+    objc_provider_fields["link_inputs"] = depset(
+        transitive = [
+            objc_provider_fields.get("link_inputs", depset([])),
+            depset(static_framework_file + dynamic_framework_file),
+        ],
+    )
+    objc_provider_fields["static_framework_file"] = depset(transitive = [
+        objc_provider_fields.get("static_framework_file", depset([])),
+        depset(static_framework_file),
+    ])
+
+    objc_provider_fields["dynamic_framework_file"] = depset(transitive = [
+        objc_provider_fields.get("dynamic_framework_file", depset([])),
+        depset(dynamic_framework_file),
+    ])
+
+    objc = apple_common.new_objc_provider(
+        **objc_provider_fields
+    )
+
+    return [
+        DefaultInfo(files = depset(dynamic_framework_dirs)),
+        objc,
+        _make_imports([depset(dynamic_framework_dirs)]),
+    ]
+
+import_middleman = rule(
```

Comment:
I'd appreciate if you could add a doc to this rule to help others in the future understand the goals, usage, expected outputs etc